Decom old nodes: CORE-7111

joe-redpanda opened the pull request:
Feature PR for allowing brokers to automatically decommission brokers which have been unavailable for a certain timeout.

Adds last seen to a new report in the nodewise health report.

Processes last seen reports to find nodes which are past the decommission timeout on a quorum of nodes.

This creates a list of auto decom candidates, one of which will be selected (the lowers node id) to be automatically decommissioned. Nothing will be submitted for auto decommission if theres is an ongoing decommission.

Completes https://redpandadata.atlassian.net/browse/CORE-7111

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes
### Improvements

* Allow nodes to automatically decommission after a certain timeout

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
config: add auto decommission configuration

Adds partition_autobalancing_node_autodecommission_timeout_sec
which is the time in seconds after which partition balancer
planner should begin decommissioning a node which is
unresponsive.

joe-redpanda committed:
partition_balancer_*: implement basic auto decom

Wires partition_autobalancing_node_autodecommission_time into partition
balancer.

This commit adds the basicmost implementation of auto decommissioning
which is based on the last seen from the perspective of the current
controller broker. This implementation will run into problems when
controller leadership changes.

In future commits, this will be changed for a coordinated approach where
the partition_balancer_planner will instead use the cluster health
report to seek the consent of a quorum of nodes before decommissioning a
broker.

joe-redpanda committed:
auto_decommission_test: add auto decom integ tests

Adds two tests.
1. smoke test: check that we can auto decom a node if it elapses the
   auto decom timeout
2. reset test: check that node restarts DO reset the timer on auto
   decommissioning

joe-redpanda committed:
health_monitor_types: add node_liveness_report

Adds a new struct to health_monitor_types: node_liveness_report.
This part of the health report will detail internode connectivity by
detailing when each node has last heard from every other node

Adds this struct to node_health_report and node_health_report_serde.

Adds build fixes needed given the above.

joe-redpanda committed:
health_monitor_backend: wire in node_status

adds node_status to health monitor backend. This will be
used in future commits to create an auto decom status report

joe-redpanda committed:
partition_balancer: implement coord auto decom

Adds the logic which will perform coordinated auto decommisson of nodes
which have elapsed their auto decom timeout.

Now, nodes will only be slated for auto decommission if a majority of
the nodes in the cluster have indicated that the node is derelict, by
sending a node_liveness_report where the node in question has been
unresponsive for longer than the autodecommission timeout.

partition_balancer_planner will not attempt to auto decom a node if
any node is currently decommmissioning

it will only attempt to decommission one node at a time instead of
decommissioning all nodes it finds to be derelict

joe-redpanda committed:
partition_balancer_autodecom_test: add unit tests

Adds unit tests for the partition_balancer_planner's logic of
determining when to auto decommisssion a node.

joe-redpanda commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -1621,15 +1640,15 @@ ss::future<> partition_balancer_planner::get_node_drain_actions(
-/// appear because rack awareness constraint is not a hard constraint, e.g. when
-/// a rack dies and we move all replicas that resided on dead nodes to live
-/// ones).
+/// appear because rack awareness constraint is not a hard constraint, e.g.
```
forgive the formatting. I don't know why this happened and I cannot convince clang format or otherwise to undo it 
: (

joe-redpanda commented on src/v/cluster/health_monitor_types.cc:
```diff
@@ -104,7 +123,12 @@ std::ostream& operator<<(std::ostream& o, const node_health_report& r) {
+  : node_health_report_serde(
+      hr.id,
+      hr.local_state,
+      {},
```
annotate

joe-redpanda commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -196,6 +197,47 @@ struct topic_status
+ * Status for the automatic decommissioning of dead nodes
+ */
+
+struct auto_decommission_status_data {
```
Serde stuff prevents named binding which is remarkably inconvenient.

My resolution was to bounce the actual data members into a pod struct

Now, I could either copy paste the parameters into the serde struct with explicit binding from the pod struct to the serde struct (bad because of code duplication)

or I could have composed the pod struct into the serde struct (fine but makes the serde fields and etc painful)

or I could do what I did and have the serde struct inherit the data struct to get this for free.
^ I chose this.

joe-redpanda commented on src/v/cluster/tests/health_bench.cc:
```diff
@@ -102,7 +102,11 @@ struct health_bench : health_report_accessor {
+                node::local_state{},
+                std::move(topics),
+                std::nullopt,
+                std::nullopt);
```
annotate

joe-redpanda commented on src/v/cluster/tests/health_monitor_bench.cc:
```diff
@@ -63,7 +63,7 @@ make_node_health_report(size_t num_topics, size_t partitions_per_topic) {
     }
 
-    return {id, local_state, std::move(topics), std::nullopt};
+    return {id, local_state, std::move(topics), std::nullopt, std::nullopt};
```
annotate

joe-redpanda commented on src/v/cluster/tests/partition_balancer_simulator_test.cc:
```diff
@@ -740,7 +740,11 @@ class partition_balancer_sim_fixture {
+                id,
+                local_state,
+                std::move(topics),
+                std::nullopt,
```
annotate

joe-redpanda commented on src/v/cluster/tests/randoms.h:
```diff
@@ -95,7 +95,8 @@ inline node_health_report random_node_health_report() {
       tests::random_chunked_vector(random_topic_status),
-      random_ds};
+      random_ds,
+      std::nullopt};
```
annotate

joe-redpanda commented on src/v/cluster/tests/serialization_rt_test.cc:
```diff
@@ -767,7 +767,8 @@ cluster::cluster_health_report random_cluster_health_report() {
           std::move(topics),
-          random_drain_status());
+          random_drain_status(),
+          /*maybe_auto_decommission_status*/ std::nullopt);
```
I should probably add a real auto decom test to this

joe-redpanda commented on src/v/cluster/controller.cc:
```diff
@@ -1007,21 +1010,22 @@ ss::future<> controller::create_cluster(
-        // * In root driven bootstrap mode, root is the only cluster member and
-        // joiners never make it into create_cluster(), so checking for
+        // cluster_uuid to appear, or for leadership change, while the
+        // leader will proceed with cluster creation once elected.
```
not sure whats happening with the comment reformats. open to suggestions

joe-redpanda commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -2160,6 +2165,122 @@ partition_balancer_planner::get_force_repair_actions(request_context& ctx) {
+       .current_config = ctx.state().get_config_version()});
+
+    ss::sstring decommission_list
+      = "248624 finishing auto_get_decommission_actions with nodes to "
```
cleanup logs and comments

joe-redpanda commented on tests/rptest/tests/auto_decommission_test.py:
```diff
@@ -0,0 +1,347 @@
+    @cluster(num_nodes=6, log_allow_list=CHAOS_LOG_ALLOW_LIST)
+    def test_decom_timer_reset(self):
+        """
+        Test that a node is automatically decommissioned when it's unresponsive
```
fix the comment

joe-redpanda commented on tests/rptest/tests/auto_decommission_test.py:
```diff
@@ -0,0 +1,347 @@
+        was_decommissioned: bool = to_decommission_node_id not in [
+            int(broker_status["node_id"]) for broker_status in broker_statuses
+        ]
+        assert (  # pyright: ignore [reportAssertAlwaysTrue]
```
IS IT?

joe-redpanda replied:
the parens are the issue, its interpreting this as a tuple which will always evaluate to true because valid tuples are truthy

joe-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -3263,6 +3263,15 @@ configuration::configuration()
+      "applies only when `partition_autobalancing_mode` is set to "
+      "`continuous`.      ",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      24h)
```
should we add another config which is enable/disable.
I don't really want to set this to 10000 years and find out this value overflows an integer in some wacky edge case

mmaslankaprv replied:
can we just set it to `0` to disable it ? Or make it an optional ?

joe-redpanda commented on tests/rptest/tests/auto_decommission_test.py:
```diff
@@ -0,0 +1,360 @@
+# Copyright 2020 Redpanda Data, Inc.
```
I'll fix the copyright in the next revision

joe-redpanda commented:
Looks like partition balancer planner test is flaking right now on what appears to be reactor starvation?
EDIT: this happens if plan_actions completes synchronously

joe-redpanda commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -362,17 +381,17 @@ void partition_balancer_planner::init_per_node_state(
+        if (_config.mode == model::partition_autobalancing_mode::continuous) {
+            // get all the nodes which are unresponsive enough to drain
+            if (time_since_last_seen > _config.node_availability_timeout_sec) {
+                vlog(
```
how did this happen?

bharathv commented on src/v/config/configuration.cc:
```diff
@@ -3274,6 +3274,15 @@ configuration::configuration()
+      "When a node is unavailable for at least this timeout duration, it "
+      "triggers Redpanda to decommission the node. This property "
+      "applies only when `partition_autobalancing_mode` is set to "
+      "`continuous`.      ",
```
nit: extra spaces

bharathv commented on src/v/config/configuration.cc:
```diff
@@ -3274,6 +3274,15 @@ configuration::configuration()
       "`continuous`.      ",
       {.needs_restart = needs_restart::no, .visibility = visibility::user},
       15min)
+  , partition_autobalancing_node_autodecommission_time(
```
decommission_time_`secs`

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -2020,20 +2039,20 @@ ss::future<> partition_balancer_planner::get_counts_rebalancing_actions(
+              // replica moves to existing reassignments. The reason is that
+              // this will bias the algorithm towards adding more moves to
+              // partitions that we already reassigned, which we want to
+              // avoid (to avoid formation of isolated replica subsets).
```
nit: can we split non logical changes to a separate commit.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a1c383f3-c52b-40d0-9e1f-68a3e9392f31-->
<!--TEST RESULTS-->
<!--f70b14ad-d66b-466c-a4f9-56f840364f4d-->
<!--TEST RESULTS-->
<!--423223b2-7487-4d46-a2da-c77f7f8e72af-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78233">build#78233</a></summary>


| test_class               | test_method                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                        |
|:-------------------------|:-----------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetForLeaderEpochTest | test_offset_for_leader_epoch | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78233#019b38a1-c510-44be-adfc-4687969a617d | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetForLeaderEpochTest&test_method=test_offset_for_leader_epoch |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78603">build#78603</a></summary>


| test_class                      | test_method              | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78603#019b951e-87d8-4a5b-b670-b94b6507f76c | FLAKY         | 18/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0526, p0=0.2840, reject_threshold=0.0100. adj_baseline=0.1497, p1=0.4062, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78731">build#78731</a></summary>


| test_class                       | test_method                     | test_arguments                                                                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                   |
|:---------------------------------|:--------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AdjacentSegmentMergingTest       | test_reupload_of_local_segments | {"acks": 1, "cloud_storage_type": 1}                                                                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdjacentSegmentMergingTest&test_method=test_reupload_of_local_segments       |
| AdminUUIDOperationsTest          | test_uuid_override              | {"force": true, "mode": 3}                                                                                                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminUUIDOperationsTest&test_method=test_uuid_override                       |
| ClusterConfigAliasTest           | test_aliasing_with_upgrade      | {"prop_set_name": "log_retention_ms", "wipe_cache": false}                                                                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterConfigAliasTest&test_method=test_aliasing_with_upgrade                |
| DataTransformsLoggingTest        | test_logs_otel                  | null                                                                                                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataTransformsLoggingTest&test_method=test_logs_otel                         |
| MountUnmountIcebergTest          | test_simple_remount             | {"cloud_storage_type": 1}                                                                                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2c-4094-963f-961e97ff3281 | FLAKY         | 7/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.2180, p0=0.3771, reject_threshold=0.0100. adj_baseline=0.5219, p1=0.1383, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                      |
| FollowerFetchingTest             | test_basic_follower_fetching    | {"fetch_from": "fetch-from-cloud-topic"}                                                                                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FollowerFetchingTest&test_method=test_basic_follower_fetching                |
| ShadowIndexingCloudRetentionTest | test_cloud_size_based_retention | {"cloud_storage_type": 1}                                                                                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowIndexingCloudRetentionTest&test_method=test_cloud_size_based_retention |
| SchemaRegistryTest               | test_schema_id_validation       | {"client_type": 1, "compression_type": "zstd", "payload_class": "com.redpanda.Payload", "protocol": 1, "subject_name_strategy": "io.confluent.kafka.serializers.subject.TopicNameStrategy", "validate_schema_id": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaRegistryTest&test_method=test_schema_id_validation                     |
| KgoVerifierSelfTest              | test_kgo_verifier_multi         | null                                                                                                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=KgoVerifierSelfTest&test_method=test_kgo_verifier_multi                      |
| TimeQueryTest                    | test_timequery                  | {"batch_cache": true, "cloud_storage": false, "spillover": false}                                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery                                     |
| CreateTopicsResponseTest         | test_create_topic_responses     | {"partition_count": -1, "replication_factor": 3}                                                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CreateTopicsResponseTest&test_method=test_create_topic_responses             |
| TxVerifierTest                   | test_all_tx_tests               | {"cloud_storage_type": 1}                                                                                                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/78731#019b9f47-8e2d-47d1-af71-8baaf08736db | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxVerifierTest&test_method=test_all_tx_tests                                 |
</details>
<!--/body-->

mmaslankaprv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -380,6 +387,13 @@ void partition_balancer_planner::init_per_node_state(
+
+                // get all nodes which are unresponsive enough to decom
+                if (time_since_last_seen > _config.decommission_timeout) {
+                    if (!ctx.decommissioning_nodes.contains(id)) {
```
nit: try_emplace ?

joe-redpanda replied:
different lists, one tracks nodes that are currently decommissioning, one tracks nodes which are decommission-able

Sounds like a rename is in order

mmaslankaprv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -459,8 +473,8 @@ ss::future<> partition_balancer_planner::init_ntp_sizes_from_health_report(
-                // assume that the "true" non-reclaimable size is the max of all
-                // replicas.
+                // assume that the "true" non-reclaimable size is the max of
+                // all replicas.
```
unrelated change ?

mmaslankaprv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -1746,9 +1764,9 @@ size_t partition_balancer_planner::calculate_full_disk_partition_move_priority(
-    // p_size/disk_capacity to be larger than 1.0.
+    // behavior in case the size is incorrectly reported, this is required
+    // as we normalize the size with disk capacity and we do not want the
+    // ration of p_size/disk_capacity to be larger than 1.0.
```
lots of unrelated formatting changes :-)

joe-redpanda replied:
Yeah see the comment above. Clang format keeps doing this whenever I edit this file. I tried to hit the file with a preemptive 'clang-format' maybe so I can have a just format change and then an actual code change, but it refused to reformat until there were code changes...

Let me know if this is too much of a review burden and I can surgery these commits to get a less irritating diff.

joe-redpanda replied:
format on save was biting me

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -727,7 +728,11 @@ ss::future<> controller::start(
       std::ref(_partition_leaders),
-      std::ref(_tp_state));
+      std::ref(_tp_state),
+      std::ref(_node_status_table),
```
can we rely on the information from the health report itself instead of pulling in another source of health information ?

joe-redpanda replied:
~~Yeah I think by the time we hit coordinated auto decom it isn't used. let me drop in a commit that un-wires it~~

thought this was a wire into pbp

No, because health monitor backend is calculating the health reports. Generation of a health report requires node status information to divine which nodes have surpassed their auto decommission timeout.

We're generating a status of delinquent nodes on every node s.t. we can check on the controller if a quorum is reporting a node as delinquent.

joe-redpanda commented:
Fixed the unrelated formatting,
Changed the name of _decommisionable_nodes to _nodes_to_decom for reader ease

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -378,6 +381,13 @@ void partition_balancer_planner::init_per_node_state(
                     model::timestamp_clock::duration>(time_since_last_seen));
 
+                // get all nodes which are unresponsive enough to decom
+                if (time_since_last_seen > _config.decommission_timeout) {
```
* should we filter out nodes under maintenance?
* should we limit to decommissioning node (across user invoked + auto decom) at a time?

joe-redpanda replied:
maintanence mode nodes get filtered out in the quorum based approach

lets chat tomorrow about 'one at a time'

bharathv replied:
sorry I missed the dicussion, was there a conclusion on this?

bharathv commented on src/v/cluster/health_monitor_backend.h:
```diff
@@ -232,6 +236,8 @@ class health_monitor_backend {
     ss::sharded<partition_leaders_table>& _partition_leaders_table;
     ss::sharded<topic_table>& _topic_table;
+    ss::sharded<node_status_table>& _node_status_table;
+    std::unique_ptr<config_i> _config;
```
naming nit: can we use a more descriptive name, config_api or something along those lines.

bharathv commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -240,9 +284,15 @@ struct node_health_report_serde
+          local_state,
+          topics,
+          drain_status,
+          maybe_auto_decommission_status);
```
serde version bumps?

bharathv commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -196,6 +197,47 @@ struct topic_status
+    config_version configuration_version{config_version_unset};
+    // a list of the node_ids which have exceeded the auto decommission timeout
+    std::vector<model::node_id> nodes_past_auto_decom_timeout;
+};
```
can we name this something orthogonal to "auto_decommission_foo" ? 

struct node_liveness_report {

...
map<node_id, ts> _last_successful_hb_ts;

};


The user can use this info for auto decomm purposes.

joe-redpanda replied:
considering

It would be nice to have some report on (controller) follower to follower connectivity, but penned the current model because I wanted this report to be 'free' in the 99.9% of cases where no nodes are derelict.

I again wish for 'sloth mail' because I'd be happier with the data exchange if it was ~every 30 min as opposed to every health report

bharathv replied:
Not sure I’m following.. I just meant we could rename it to match what it actually means, instead of how it’s being used.

Also, separate question: where is `nodes_past_auto_decom_timeout` actually set? I’m not seeing anywhere it gets modified, or maybe github is messing with me.

joe-redpanda replied:
Yeah, I'm saying that it currently only has nodes past auto decom timeout because the field is optional. we only populate it when there is a node past auto decom timeout.

It is fair, though, to change this to something like map<node, timestamp> and always include the field with entries for all nodes.

^my concern is that this will be on all health reports (right now we only pay network cost if theres a delinquent node)
Is this useful enough to afford this?

bharathv commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -1053,6 +1057,76 @@ health_monitor_backend::collect_topic_status() {
+      maybe_config_version,
+      "current implementation of config_i::get_config should never return an "
+      "error");
+    auto config_version = *maybe_config_version;
```
can you please clarify the purpose of this config version?

joe-redpanda replied:
current model is each node reports when its sees another node as down for "too long"

config version is for making sure every node has the same definition of 'too long'

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -2149,6 +2156,124 @@ partition_balancer_planner::get_force_repair_actions(request_context& ctx) {
+    // 2. only reports that have a value
+    // 3. only reports that have the current configuration version
+    //    - this checks that we are working with the current decommission
+    //      timeout
```
oh I see the purpose of config_version here.. I'm wondering if this is really needed. Can each broker just report the time_since_last_hb of every other broker and the controller makes a decision based based on the (latest) config version it sees (eliminates all this config_i propagation).

joe-redpanda replied:
The case bouncing around my brain was something like
a node goes down for a while
customer turns off auto decom
customer turns it back on
node returns and starts voting on who to vote off the island

I feel like config_version was the best answer to this ('hello old node, please catch up on current events and then we'll talk about decoms') and config_i did a good job of fire-walling the complexity of config_backend

bharathv replied:
> node returns and starts voting on who to vote off the island

hows that possible? Since this only happens on the coordinator leader, it would've caught up with "config turned off" update before starting the tick?

This seems like a pretty rare case anyway. It’s the kind of config you set once and then forget about (and we probably only use it internally for cloud). I think it’s fine to just ignore it..the upside is the code ends up being a lot simpler.

joe-redpanda replied:
a node which was disconnected reconnects. It'll stop showing up as unresponsive in node status and start sending health reports, possibly before getting the append entries that contains new config.

Lets chat in scrum on pros and cons of config_i
I wrote it because I thought 'it cant possibly be this hard to get the cluster config version' when looking at all the dependencies of config_backend.

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -2206,12 +2331,18 @@ partition_balancer_planner::plan_actions(
         co_return result;
     }
 
+    if (ctx.config().mode == model::partition_autobalancing_mode::continuous) {
+        get_auto_decommission_actions(ctx, health_report);
+    }
```
I"m thinking if this should happen before we populate the per node states. Because here we issue a decomm but that is not reflected in the node states and later we may schedule some moves into that node which is wasteful.

joe-redpanda replied:
That's wise. I'll add this to the allocation constraints

bharathv replied:
A simpler and non-invasive fix is to just restart the tick. In the next tick this node will be automatically populated in the list of decommissioning nodes and the usual decomm rules apply.

joe-redpanda replied:
I actually don't think this is an issue.
By the time a node is unresponsive long enough to get decommed, it should already have been preemptively drained.

joe-redpanda replied:
Regardless, if we did something like this, we would either have to do some invasive changes or invoke weird behavior. either
1. we early exit the planning tick and let the next tick be called normally (in 30s or so)
2. we change the behavior to start a decom, do a linearizable barrier, and replan

bharathv reviewed (commented):
didn't look at tests yet, flushing out remaining comments.

joe-redpanda committed:
utils/to_string: add a formatter for flat_hash_set

Adds a convenience formatter for flat hash set s.t. it can be easily
logged.

joe-redpanda committed:
health_monitor_backend: populate node_liveness

Adds population of node liveness report from the node_status_table.
This will iterate over all cluster members, fetching their last seen
from the node status table (if present) and populating the
node_liveness_report with the results.

joe-redpanda committed:
partition_balancer_planner: add auto decom time...

to config

Adds node_autodecommission_timeout to
parition_balancer_planner::planner_config such that it can later be used
in auto decommission logic.

Adds necessary build fixes for this.

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -3277,6 +3274,15 @@ configuration::configuration()
+      "When a node is unavailable for at least this timeout duration, it "
+      "triggers Redpanda to decommission the node. This property "
+      "applies only when `partition_autobalancing_mode` is set to "
+      "`continuous`",
```
```suggestion
      "`continuous`.",
```

bharathv commented on src/v/utils/to_string.h:
```diff
@@ -61,6 +62,22 @@ struct fmt::formatter<ss::chunked_fifo<T, chunk_size>> {
+        first = false;
+    }
+    o << "}";
+    return o;
```
fmt::join(set, ", ") ?

joe-redpanda replied:
awesome, yeah I'll do that

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -378,6 +381,13 @@ void partition_balancer_planner::init_per_node_state(
 
+                // get all nodes which are unresponsive enough to decom
+                if (time_since_last_seen > _config.decommission_timeout) {
+                    if (!ctx.decommissioning_nodes.contains(id)) {
```
nit: L386 is redundant?

joe-redpanda replied:
yep, looks like this was a rebase mistake

joe-redpanda replied:
ah, no, the flow of this pr is config -> basicmost smoke test implementation -> scaffolding for coordinated decommission -> implementation of coordinated decommission -> coordinated decommissioning

this code block gets removed here:
https://github.com/redpanda-data/redpanda/pull/28946/commits/b19ab425a138e786b8a8fe2ba479b692ff564deb#r2673576087

when simple decom gets swapped to coordinated decom

bharathv commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -196,6 +198,41 @@ struct topic_status
+ * Status for the automatic decommissioning of dead nodes
+ */
+
+struct node_liveness_report_data {
```
curious why this split.

joe-redpanda replied:
allows named binding, it made more sense when there were multiple members in auto_decommission_report prior so I can take or leave the split for this use case, will probably use in the future though because I hate constructors with a bunch of positional parameters

bharathv commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -196,6 +198,41 @@ struct topic_status
+
+struct node_liveness_report_data {
+    // map between a given node id and the time at which it was last heard from
+    absl::flat_hash_map<model::node_id, rpc::clock_type::duration>
```
nit: comment says "time at which.." while the value_type is a duration,

bharathv commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -2149,6 +2159,193 @@ partition_balancer_planner::get_force_repair_actions(request_context& ctx) {
+
+              vlog(
+                clusterlog.info,
+                "248624 map size was {}",
```
remove

joe-redpanda replied:
whoops, thank you

joe-redpanda commented:
Rebase onto dev to cut my build times, no relevant changes

joe-redpanda commented on src/v/cluster/partition_balancer_planner.cc:
```diff
@@ -381,13 +398,6 @@ void partition_balancer_planner::init_per_node_state(
                   - std::chrono::duration_cast<
                     model::timestamp_clock::duration>(time_since_last_seen));
 
-                // get all nodes which are unresponsive enough to decom
```
anchor comment

dotnwat reviewed (commented):
@joe-redpanda please update this PR cover letter to explain what the PR is doing, give it a meaningful title, and link to any related JIRA tickets.

david-yu commented:
@joe-redpanda What are the chances of us being able to backport this to 25.3, 25.2, and 25.1? The reason we ask as is we can replace the logic we have entirely in our decommission controller as per @andrewstucki which would be huge for us in terms maintenance as this approach is much cleaner.

joe-redpanda commented:
> @joe-redpanda What are the chances of us being able to backport this to 25.3, 25.2, and 25.1? The reason we ask as is we can replace the logic we have entirely in our decommission controller as per @andrewstucki which would be huge for us in terms maintenance as this approach is much cleaner.

Backports are almost exclusively reserved for high impact bug fixes. I would say backporting a feature, and especially one of this magnitude is not a good idea.

We should also give this time to bake to crop up any unexpected behavior. I would like to avoid a situation where automatic decommission somehow triggers prematurely, leading to a kubernetes cluster with a pod that has been kicked out of the underlying redpanda cluster, because I imagine there are no automatic mitigation pathways that will correct this.

david-yu commented:
Ok thank you I'll take your lead on this one. Not urgent but we thought it was worth asking.
